Wikipedia/Code/WMFAppViewController.m

```diff
@@ -1927,6 +1975,113 @@ - (void)entriesLimitReachedWithNotification:(NSNotification *)notification {
     }
 }
 
+#pragma mark - Remote Notifications
+
+- (void)startRemoteNotificationsController {
+    if (self.dataStore.wikidataDescriptionEditingController.madeAuthorizedWikidataDescriptionEdit) {
+        [self.dataStore.remoteNotificationsController start];
+    }
+}
+
+- (void)remoteNotificationsModelDidChange:(NSNotification *)note {
+    RemoteNotificationsModelChangeResponseCoordinator *responseCoordinator = (RemoteNotificationsModelChangeResponseCoordinator *)note.object;
+    RemoteNotificationsModelChange *modelChange = (RemoteNotificationsModelChange *)responseCoordinator.modelChange;
+    NSDictionary<NSNumber *, NSArray<RemoteNotification *> *> *notificationsGroupedByCategoryNumber = (NSDictionary<NSNumber *, NSArray<RemoteNotification *> *> *)modelChange.notificationsGroupedByCategoryNumber;
+    assert(responseCoordinator);
+    assert(modelChange);
+    if (modelChange.type != RemoteNotificationsModelChangeTypeAddedNewNotifications) {
+        return;
+    }
+    NSNumber *editRevertedCategoryNumber = [NSNumber numberWithInt:RemoteNotificationCategoryEditReverted];
+    NSArray<RemoteNotification *> *editRevertedNotifications = notificationsGroupedByCategoryNumber[editRevertedCategoryNumber];
+    if (editRevertedNotifications.count == 0) {
+        return;
+    }
+    // TODO: How do we handle multiple notifications?
+    NSString *alertMessage;
+    NSString *notificationBody;
+    NSString *notificationTitle;
+    NSMutableArray<WMFArticle *> *articles = [NSMutableArray new];
+    NSMutableArray<NSString *> *articleKeys = [NSMutableArray new];
+    NSMutableArray<RemoteNotification *> *filteredNotifications = [NSMutableArray new];
+    NSMutableArray<NSString *> *filteredNotificationsIDs = [NSMutableArray new];
+
+    for (RemoteNotification *editRevertedNotification in editRevertedNotifications) {
+        WMFArticle *article = [self.dataStore fetchArticleWithWikidataID:editRevertedNotification.affectedPageID];
+        if (article) {
+            // Hide notifications without article context
+            [filteredNotifications addObject:editRevertedNotification];
+            // We don't create notifications without ids so the objects that we're adding won't be nil
+            [filteredNotificationsIDs addObject:editRevertedNotification.id];
+            [articles addObject:article];
+            [articleKeys addObject:article.key];
+        }
```

Comment:
For our current use case, if we don't have article context for a given notification, my assumption is that we probably don't want to show it to the user - if that's correct, I'm not sure what to do with those notifications. Delete them and then re-save them when we fetch them from the server, check if there's article context again, repeat? Or maybe it'd be better to check for the article context when we fetch the notifications, instead of waiting until we're about to show them to the user